### PR TITLE
gh-59705: Export threading.Thread() names to the OS

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Call pthread_setname_np() et al. with the name argument passed to
threading.Thread().

https://bugs.python.org/issue15500

* Issue: gh-59705

### CODE REVIEW

**Comment by ZackerySpytz** [comment]
Unfortunately, the string passed to `pthread_setname_np()` must be less than 16 chars on Linux, so the name may be truncated.

**Comment by debauer** [comment]
Will this ever happen?! The Issue has this year the 10th anniversary. In the Issue @eryksun also provides Code for Windows... so let this happen.

File: Lib/threading.py [hunk 1]
```diff
@@ -924,6 +929,8 @@ def _bootstrap_inner(self):
             self._set_tstate_lock()
             if _HAVE_THREAD_NATIVE_ID:
                 self._set_native_id()
+            if _HAVE_SET_THREAD_NAME:
+                _set_thread_name(self.name)
```
Comment (pitrou) [inline Lib/threading.py:933]: This should probably be more sophisticated, e.g. `f"{sys.executable} [thread {self.name!r}]"`

File: Python/thread_pthread.h [hunk 2]
```diff
@@ -343,6 +344,37 @@ PyThread_get_thread_native_id(void)
...
+    size_t len = Py_ARRAY_LENGTH(buf) - 1;
+    strncpy(buf, name, len);
+    buf[len] = '\0';
+#ifndef __APPLE__
+    unsigned long ident = PyThread_get_thread_ident();
+#endif
+    int ret = 0;
+#ifdef __APPLE__
+    // On macOS, pthread_setname_np() can only set the calling thread's name.
+    ret = pthread_setname_np(buf);
```
Comment (encukou) [inline Python/thread_pthread.h:365]: It seems (from the man page) that macOS doesn't have the 16-character limitation. Could we pass the original buffer here?

File: Lib/threading.py (hunk 1 above)
Comment (encukou) [reply Lib/threading.py:933]: That wouldn't work well for platforms with the 16-character limitation.

**Review by encukou — COMMENTED** [review]
This needs tests, which in turn will need a getter function -- at least a private one.

**Comment by vstinner** [comment]
New attempt to implement this feature: https://github.com/python/cpython/pull/127338

**Comment by vstinner** [comment]
https://github.com/python/cpython/pull/127338 was merged instead. Thanks for your contribution anyway!

**Pull request closed without merging**